hivemq-edge/src/frontend/src/components/rjsf/BatchSubscription/components/ColumnMatcherStep.tsx

```diff
@@ -1,7 +1,128 @@
+    if (!header) return []
+    //  TODO[NVL] deal with duplicate headers (seems to be adding _number)
+    //  TODO[NVL] deal with no header!
+    return Object.keys(header).map((e) => ({ value: e, label: e }))
+  }, [worksheet])
+
+  type FormValues = {
+    mapping: ColumnMappingData[]
+  }
+
+  const {
+    register,
+    control,
+    trigger,
+    getValues,
+    formState: { errors, isValid },
+  } = useForm<FormValues>({
+    defaultValues: {
+      mapping: subscriptions.map((subscription) => {
+        const autoMatch = findMatch(subscription, columns)
+
+        return { column: autoMatch || '', subscription: subscription.label }
+      }),
+    },
+    mode: 'onChange',
+    reValidateMode: 'onChange',
+  })
+  const { fields } = useFieldArray({
+    name: 'mapping',
+    control,
+  })
+
+  useEffect(() => {
+    onContinue({ mapping: undefined })
+    trigger()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  useEffect(() => {
+    if (isValid) {
+      const { mapping } = getValues()
+      onContinue({ mapping: mapping })
+    }
+  }, [getValues, isValid, onContinue])
+
+  return (
+    <>
+      <chakra.form id="batch-mapping-form">
+        {fields.map((field, index) => {
+          return (
+            <FormControl key={field.id} isInvalid={!!errors?.mapping?.[index]?.column} p={4}>
+              <HStack>
+                <Box flex={1}>
+                  <Controller
+                    name={`mapping.${index}.column`}
+                    control={control}
+                    rules={{ required: t('rjsf.batchUpload.columnMapping.validation.required') }}
+                    render={({ field: { value, onChange, ...rest } }) => {
+                      return (
+                        <Select<ColumnOption>
+                          // TODO[modal-portal] Feels like a hack. Use a ref [?]
+                          menuPortalTarget={document.getElementById('chakra-modal-array-field-batch') as HTMLElement}
+                          id={`mapping.${index}.column`}
+                          instanceId={`mapping.${index}.column`}
+                          aria-label={t('rjsf.batchUpload.columnMapping.column.ariaLabel')}
+                          {...rest}
+                          onChange={(e) => onChange(e?.label)}
+                          value={{ label: value, value }}
+                          options={columns}
+                        />
+                      )
+                    }}
+                  ></Controller>
+                </Box>
+                <LuChevronsRight />
+                <Box flex={1}>
+                  <Input
+                    aria-label={t('rjsf.batchUpload.columnMapping.subscription.ariaLabel')}
+                    readOnly
+                    {...register(`mapping.${index}.subscription`, {
+                      required: true,
+                    })}
+                    className={errors?.mapping?.[index]?.subscription ? 'error' : ''}
```

Comment:
issue(non-blocking): register properties might overwrite your manual aria-label and readOnly properties

```suggestion
                    {...register(`mapping.${index}.subscription`, {
                      required: true,
                    })}
                    aria-label={t('rjsf.batchUpload.columnMapping.subscription.ariaLabel')}
                    readOnly
                    className={errors?.mapping?.[index]?.subscription ? 'error' : ''}
```